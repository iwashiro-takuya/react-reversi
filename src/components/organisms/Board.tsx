import React from 'react';
import { useDispatch } from 'react-redux';
import styled from 'styled-components';
import { SqiuareState } from '../../types';
import { addSquares } from '../../actions';
import Square from '../molecules/Square';

import Const from '../../const';

const { Color } = Const;

interface BoardProps {
  onSideSquares: number;
}

interface Wrapper {
  side: number;
}

const Board: React.FC<BoardProps> = ({ onSideSquares }) => {
  const dispatch = useDispatch();

  const totalSquaresCounts = Array.from(new Array(onSideSquares ** 2).keys());

  const initializeSquaresPosition = totalSquaresCounts.map(
    (i: number): SqiuareState => {
      return {
        column: Math.floor(i / onSideSquares),
        row: i % onSideSquares,
        val: 0,
      };
    },
  );

  dispatch(addSquares(initializeSquaresPosition));

  return (
    <Wrapper side={onSideSquares * 80}>
      {totalSquaresCounts.map((i: number) => (
        <Square key={i} piece="block" />
      ))}
    </Wrapper>
  );
};

const Wrapper = styled.div<Wrapper>`
  display: flex;
  flex-wrap: wrap;
  width: ${(props) => props.side}px;
  height: ${(props) => props.side}px;
  margin: 0 auto;
  border: 1px solid ${Color.BD_BLACK};
`;

export default Board;
