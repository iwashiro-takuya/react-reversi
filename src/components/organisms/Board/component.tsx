import React from 'react';
import styled from 'styled-components';
import useBoard from 'src/components/organisms/Board/hook';
import Square from 'src/components/atoms/Square/component';
import Piece from 'src/components/atoms/Piece/component';
import Const from 'src/const';

const { Size, Color } = Const;

export interface BoardProp {
  dataCy: string;
}
const Board: React.FC<BoardProp> = ({ dataCy }) => {
  const {
    boardSquaresArray,
    sideSquaresCount,
    hasReversiblePiece,
    reverseSquare,
    hasPlacedPiece,
  } = useBoard();

  return (
    <StyledBoard size={sideSquaresCount * Size.SQUARE_SIZE} data-cy={dataCy}>
      {boardSquaresArray.map((square: SquareState) => {
        return (
          <Square key={square.id} dataCy={`square-${square.id}`}>
            {hasReversiblePiece(square.id) && (
              <Piece
                playerVal={square.val}
                onclick={() => {
                  reverseSquare(square.id);
                }}
                dataCy="clickable"
              />
            )}
            {hasPlacedPiece(square.id) && <Piece playerVal={square.val} />}
          </Square>
        );
      })}
    </StyledBoard>
  );
};

interface StyledBoardProp {
  size: number;
}
const StyledBoard = styled.div<StyledBoardProp>`
  display: flex;
  flex-wrap: wrap;
  width: ${(props) => props.size}px;
  height: ${(props) => props.size}px;
  margin-bottom: 16px;
  border: 1px solid ${Color.BD_BLACK};
`;

export default Board;
