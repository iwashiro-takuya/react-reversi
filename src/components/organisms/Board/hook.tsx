/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { useSelector, useDispatch } from 'react-redux';
import { updateBoardState, updateCurrentPlayer } from 'src/redux/modules/game';
import Const from 'src/const';

const { PlayerVal } = Const;

const useBoard = (): {
  boardSquaresArray: BoardState;
  sideSquaresCount: number;
  hasReversiblePiece: (squareCount: number) => boolean;
  reverseSquare: (squareCount: number) => void;
  hasPlacedPiece: (squareCount: number) => boolean;
} => {
  const dispatch = useDispatch();
  const {
    sideSquaresCount,
    boardState: boardSquaresArray,
    currentPlayer,
  } = useSelector((state: StoreState) => state.game);

  // Method to get the square to turn over from the passed array.....
  const getShouldReverseSquareArray = (
    baseSquare: SquareState,
    board: BoardState,
    sideSquares: number,
    playerVal: UnionValType<typeof PlayerVal>,
  ): SquareState[] => {
    const aroundSquareArrays: Array<SquareState[]> = [];

    // Get the array of squares from the starting square to the left of the board.
    aroundSquareArrays.push(
      board
        .filter((square: SquareState) => {
          return !!(
            square.row === baseSquare.row && square.column < baseSquare.column
          );
        })
        .reverse(),
    ); // Reverse to use the clicked square as a starting point

    // Get the array of squares on the right side of the board from the starting square
    aroundSquareArrays.push(
      board.filter((square: SquareState) => {
        return !!(
          square.row === baseSquare.row && square.column > baseSquare.column
        );
      }),
    );

    // Get an array of squares from the starting square to the top of the board
    aroundSquareArrays.push(
      board
        .filter((square: SquareState) => {
          return !!(
            square.row < baseSquare.row && square.column === baseSquare.column
          );
        })
        .reverse(),
    ); // Get an array of squares from the starting square to the top of the board

    // Obtain an array of squares from the starting square to the bottom of the board
    aroundSquareArrays.push(
      board.filter((square: SquareState) => {
        return !!(
          square.row > baseSquare.row && square.column === baseSquare.column
        );
      }),
    );

    // Obtain an array of squares from the starting square to the top left corner of the board
    const upperLeftDiagonalSideSquareArray: SquareState[] = [];
    for (let count = 1; count < sideSquares; count += 1) {
      const squareId: number = baseSquare.id - sideSquares * count - count;
      if (squareId >= 0 && squareId < board.length) {
        upperLeftDiagonalSideSquareArray.push(board[squareId]);
      }
    }
    aroundSquareArrays.push(upperLeftDiagonalSideSquareArray);

    // Obtain an array of squares from the starting square to the top right corner of the board
    const upperRightDiagonalSideSquareArray: SquareState[] = [];
    for (let count = 1; count < sideSquares; count += 1) {
      const squareId: number = baseSquare.id - sideSquares * count + count;
      if (squareId > 0 && squareId < board.length) {
        upperRightDiagonalSideSquareArray.push(board[squareId]);
      }
    }
    aroundSquareArrays.push(upperRightDiagonalSideSquareArray);

    // Obtain an array of squares from the starting square to the bottom left corner of the board

    const lowerLeftDiagonalSideSquareArray: SquareState[] = [];
    for (let count = 1; count < sideSquares; count += 1) {
      const squareId: number = baseSquare.id + sideSquares * count - count;
      if (squareId >= 0 && squareId < board.length - 1) {
        lowerLeftDiagonalSideSquareArray.push(board[squareId]);
      }
    }
    aroundSquareArrays.push(lowerLeftDiagonalSideSquareArray);

    // Obtain an array of squares from the starting square to the bottom right corner of the board
    const lowerRightDiagonalSideSquareArray: SquareState[] = [];
    for (let count = 1; count < sideSquares; count += 1) {
      const squareId: number = baseSquare.id + sideSquares * count + count;
      if (squareId >= 0 && squareId < board.length) {
        lowerRightDiagonalSideSquareArray.push(board[squareId]);
      }
    }
    aroundSquareArrays.push(lowerRightDiagonalSideSquareArray);

    const shouldReverseSquareArray: SquareState[] = [];
    aroundSquareArrays.forEach((squareArray) => {
      const emptySquareIndex: number = squareArray.findIndex(
        (square) => square.val === 0,
      );
      if (emptySquareIndex !== -1) {
        squareArray.splice(emptySquareIndex, squareArray.length);
      }

      const endpointSquareIndex: number = squareArray.findIndex(
        (square) => square.val === playerVal,
      );
      squareArray.splice(Math.max(0, endpointSquareIndex), squareArray.length);

      shouldReverseSquareArray.push(...squareArray);
    });

    return shouldReverseSquareArray;
  };

  // Method to check if there is a stone that can be turned over
  const hasReversiblePiece = (squareCount: number): boolean => {
    const baseSquare: SquareState = boardSquaresArray[squareCount];
    const shouldReverseSquareArray: SquareState[] = getShouldReverseSquareArray(
      baseSquare,
      boardSquaresArray,
      sideSquaresCount,
      currentPlayer,
    );

    return !!(baseSquare.val === 0) && !!(shouldReverseSquareArray.length > 0);
  };

  // A method to check if a stone has already been placed
  const hasPlacedPiece = (squareCount: number): boolean => {
    const baseSquare: SquareState = boardSquaresArray[squareCount];

    return !!(baseSquare.val !== 0);
  };

  // The method turns over the stone that was trapped when the stone was placed and replaces the player
  const reverseSquare = (squareCount: number) => {
    const stateCopy: typeof boardSquaresArray = boardSquaresArray.slice();
    const clickedSquare: SquareState = stateCopy[squareCount];

    // The method turns over the stone that was trapped when the stone was placed and replaces the player
    const shouldReverseSquareArray: SquareState[] = getShouldReverseSquareArray(
      clickedSquare,
      stateCopy,
      sideSquaresCount,
      currentPlayer,
    );

    // Batch change the values of cells to be flipped
    shouldReverseSquareArray.forEach((square) => {
      stateCopy[square.id].val = currentPlayer;
    });

    clickedSquare.val = currentPlayer;

    const nextPlayer =
      currentPlayer === PlayerVal.BLACK ? PlayerVal.WHITE : PlayerVal.BLACK;
    dispatch(updateBoardState(stateCopy));
    dispatch(updateCurrentPlayer(nextPlayer));
  };

  return {
    boardSquaresArray,
    sideSquaresCount,
    hasReversiblePiece,
    reverseSquare,
    hasPlacedPiece,
  };
};

export default useBoard;
