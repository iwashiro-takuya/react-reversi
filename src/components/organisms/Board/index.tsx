import React from 'react';
import styled from 'styled-components';
import Square from 'src/components/atoms/square/index';
import Piece from 'src/components/atoms/piece/index';
import Const from 'src/const';
import useBoard from './hook';

const { Size, Color } = Const;

export interface BoardProps {
  dataCy: string;
}
const Board: React.FC<BoardProps> = ({ dataCy }) => {
  const {
    board,
    sideSquaresCount,
    hasCanBeTurnOverPieces,
    hasPlacedPiece,
    placePiece,
  } = useBoard();

  return (
    <Wrapper size={sideSquaresCount * Size.SQUARE_SIZE} data-cy={dataCy}>
      {board.map((square: Square) => {
        return (
          <Square key={square.key} dataCy={`square-${square.key}`}>
            {hasCanBeTurnOverPieces(square) && (
              <Piece
                playerVal={square.val}
                onclick={() => {
                  placePiece(square);
                }}
                dataCy="clickable"
              />
            )}
            {hasPlacedPiece(square) && <Piece playerVal={square.val} />}
          </Square>
        );
      })}
    </Wrapper>
  );
};

interface StyledBoardProps {
  size: number;
}
const Wrapper = styled.div<StyledBoardProps>`
  display: flex;
  flex-wrap: wrap;
  width: ${(props) => props.size}px;
  height: ${(props) => props.size}px;
  margin-bottom: 16px;
  border: 1px solid ${Color.BD_BLACK};
`;

export default Board;
