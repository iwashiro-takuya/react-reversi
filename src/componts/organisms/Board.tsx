import React from 'react';
import styled from 'styled-components';
import Const from '../../const';

import Square from '../molecules/Square';

const { Color } = Const;

interface BoardProps {
  squares: number;
}

const Board: React.FC<BoardProps> = ({ squares }) => (
  <Wrapper>
    {Array.from(new Array(squares).keys()).map((i: number) => (
      <Square key={i + 1} />
    ))}
  </Wrapper>
);

const Wrapper = styled.div`
  border: 1px solid ${Color.BD_BLACK};
  display: flex;
  flex-wrap: wrap;
  height: 640px;
  margin: 0 auto;
  width: 640px;
`;

export default Board;
