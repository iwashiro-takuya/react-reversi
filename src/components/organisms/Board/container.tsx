import React from 'react';
import useBoard from 'src/components/organisms/Board/hook';
import Board, { BoardProp } from 'src/components/organisms/Board/component';
import Square from 'src/components/atoms/Square/component';
import Piece from 'src/components/atoms/Piece/container';

const EnhancedBoard: React.FC<BoardProp> = ({ onSideSquares, dataCy }) => {
  const {
    boardState,
    hasReversiblePiece,
    reverseSquare,
    hasPlacedPiece,
  } = useBoard(onSideSquares);

  return (
    <Board onSideSquares={onSideSquares} dataCy={dataCy}>
      {boardState.map((square: SquareType) => {
        return (
          <Square key={square.id} dataCy={`square-${square.id}`}>
            {hasReversiblePiece(square.id) && (
              <Piece
                playerVal={square.val}
                onclick={() => {
                  reverseSquare(square.id);
                }}
                dataCy="clickable"
              />
            )}
            {hasPlacedPiece(square.id) && <Piece playerVal={square.val} />}
          </Square>
        );
      })}
    </Board>
  );
};

export default EnhancedBoard;
