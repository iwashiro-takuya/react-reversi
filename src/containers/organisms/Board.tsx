import React from 'react';
import useBoard from '../../hooks/use-board';
import Board from '../../components/organisms/Board';
import Square from '../../components/atoms/Square';
import Piece from '../../components/atoms/Piece';

interface BoardProp {
  onSideSquares: number;
}
const EnhancedBoard: React.FC<BoardProp> = ({ onSideSquares }) => {
  const {
    squaresCountsArray,
    boardState,
    hasReversiblePiece,
    reverseSquare,
    hasPlacedPiece,
  } = useBoard(onSideSquares);

  return (
    <Board onSideSquares={onSideSquares}>
      {squaresCountsArray.map((squareCount: number) => (
        <Square key={boardState[squareCount].id}>
          {hasReversiblePiece(squareCount) && (
            <Piece
              playerVal={boardState[squareCount].val}
              onclick={() => {
                reverseSquare(squareCount);
              }}
            />
          )}

          {hasPlacedPiece(squareCount) && (
            <Piece playerVal={boardState[squareCount].val} />
          )}
        </Square>
      ))}
    </Board>
  );
};

export default EnhancedBoard;
