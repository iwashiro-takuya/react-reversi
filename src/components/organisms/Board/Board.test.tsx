import React from 'react';
import renderer from 'react-test-renderer';
import Board from './container';

it('Game snapshot testing', () => {
  const component = renderer.create(<Board onSideSquares={8} dataCy="board" />);
  expect(component).toMatchSnapshot();
});
