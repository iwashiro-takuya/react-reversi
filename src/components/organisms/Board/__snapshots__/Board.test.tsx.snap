// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board unit test snapshot 1`] = `
<div
  className="sc-bdnylx jtyzrT"
  data-cy="board"
  size={512}
>
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-0"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-1"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-2"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-3"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-4"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-5"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-6"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-7"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-8"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-9"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-10"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-11"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-12"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-13"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-14"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-15"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-16"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-17"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-18"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-19"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-20"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-21"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-22"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-23"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-24"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-25"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-26"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-27"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-28"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-29"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-30"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-31"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-32"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-33"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-34"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-35"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-36"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-37"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-38"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-39"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-40"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-41"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-42"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-43"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-44"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-45"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-46"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-47"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-48"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-49"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-50"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-51"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-52"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-53"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-54"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-55"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-56"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-57"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-58"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-59"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-60"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-61"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-62"
  />
  <div
    className="sc-gtssRu iFJSiu"
    data-cy="square-63"
  />
</div>
`;
